### PR TITLE
Hide Strong Session Affinity code behind a flag (fully) 

### PR DESCRIPTION
**Pull request by code-elinka**
Use `enableStrongSessionAffinity` flag to close access to Strong Session Affinity feature if not needed. 

This change includes: 
* @mmamczur commit 
* my changes to use excisting flag for SSA

Both of commits will be squashed together just before merging. 

That approach is called feature flagging. 
I found an article with more info about it: https://arxiv.org/pdf/1907.06157.pdf

### CODE REVIEW

File: pkg/loadbalancers/l4netlb.go [hunk 1]
```diff
@@ -303,25 +310,34 @@ func (l4netlb *L4NetLB) provideBackendService(syncResult *L4NetLBSyncResult, hcL
...
-		syncResult.GCEResourceInError = annotations.BackendServiceResource
-		syncResult.Error = err
-	}
 	var bs *composite.BackendService
-	bs, err = l4netlb.backendPool.EnsureL4BackendService(bsName, hcLink, string(protocol), string(l4netlb.Service.Spec.SessionAffinity), string(cloud.SchemeExternal), l4netlb.NamespacedName, *network.DefaultNetwork(l4netlb.cloud), connectionTrackingPolicy)
-	if err != nil {
-		if utils.IsUnsupportedFeatureError(err, strongSessionAffinityFeatureName) {
+	var err error
+	// Gate off SSA code if it's not enabled with a flag
+	if l4netlb.enableStrongSessionAffinity {
```
Comment (mmamczur) [inline pkg/loadbalancers/l4netlb.go:316]: I think it would be sufficient to have the code to get connectionTrackingPolicy gated and just pass the flag false if it's not enabled

Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go:317]: Done, thank you

File: pkg/loadbalancers/l4netlb.go [hunk 2]
```diff
@@ -411,8 +427,14 @@ func (l4netlb *L4NetLB) ensureIPv4NodesFirewall(nodeNames []string, ipAddress st
 // It is health check, firewall rules and backend service
 func (l4netlb *L4NetLB) EnsureLoadBalancerDeleted(svc *corev1.Service) *L4NetLBSyncResult {
 	isMultinetService := l4netlb.networkResolver.IsMultinetService(svc)
-	enabledStrongSessionAffinity := annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
-	result := NewL4SyncResult(SyncTypeDelete, svc, isMultinetService, enabledStrongSessionAffinity)
+	var result *L4NetLBSyncResult
+	// Gate off the SSA feature if not enabled
+	if l4netlb.enableStrongSessionAffinity {
+		enabledStrongSessionAffinity := annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
```
[the remark this answers is not in the document]
Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go]: Done

File: pkg/backends/backends_test.go [hunk 3]
```diff
@@ -228,6 +247,19 @@ func TestBackendSvcEqual(t *testing.T) {
 			},
 			wantEqual: false,
 		},
+		{
+			desc:                      "Test to check if Golang default timeout (0) will case a change",
```
Comment (mmamczur) [inline pkg/backends/backends_test.go:251]: I'd name this if "empty struct will cause..."

File: pkg/backends/backends_test.go [hunk 4]
```diff
@@ -322,11 +356,38 @@ func TestBackendSvcEqual(t *testing.T) {
 			},
 			wantEqual: true,
 		},
+		{
+			desc:                      "Test with ignoring connection tracking",
```
Comment (mmamczur) [inline pkg/backends/backends_test.go:360]: let's have a case when it should be ignored but one of them is nil

File: pkg/backends/regional_ig_linker_test.go [hunk 5]
```diff
@@ -234,7 +234,7 @@ func createBackendService(t *testing.T, sp utils.ServicePort, backendPool *Backe
 	serviceAffinityNone := string(apiv1.ServiceAffinityNone)
 	schemeExternal := string(cloud.SchemeExternal)
 	defaultNetworkInfo := network.NetworkInfo{IsDefault: true}
-	if _, err := backendPool.EnsureL4BackendService(sp.BackendName(), hcLink, protocol, serviceAffinityNone, schemeExternal, namespacedName, defaultNetworkInfo, &composite.BackendServiceConnectionTrackingPolicy{}); err != nil {
+	if _, err := backendPool.EnsureL4BackendService(sp.BackendName(), hcLink, protocol, serviceAffinityNone, schemeExternal, namespacedName, defaultNetworkInfo, false, &composite.BackendServiceConnectionTrackingPolicy{}); err != nil {
```
Comment (mmamczur) [inline pkg/backends/regional_ig_linker_test.go:237]: it doesn't change much it would be best to pass nil everywhere where this TrackingPolicy is not needed

File: pkg/loadbalancers/l4netlb.go [hunk 6]
```diff
@@ -185,7 +185,12 @@ func (l4netlb *L4NetLB) checkStrongSessionAffinityRequirements() *utils.UserErro
 // This function does not link instances to Backend Service.
 func (l4netlb *L4NetLB) EnsureFrontend(nodeNames []string, svc *corev1.Service) *L4NetLBSyncResult {
 	isMultinetService := l4netlb.networkResolver.IsMultinetService(svc)
-	enabledStrongSessionAffinity := annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
+	var enabledStrongSessionAffinity bool
+	if l4netlb.enableStrongSessionAffinity {
```
Comment (mmamczur) [inline pkg/loadbalancers/l4netlb.go:189]: you could use similar code to the one in EnsureLoadBalancerDeleted()

Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go]: thank you

File: pkg/backends/regional_ig_linker_test.go (hunk 5 above)
Comment (code-elinka) [reply pkg/backends/regional_ig_linker_test.go]: changed, ptal

File: pkg/backends/backends_test.go (hunk 4 above)
Comment (code-elinka) [reply pkg/backends/backends_test.go:373]: done, ptal :)

File: pkg/backends/backends_test.go (hunk 3 above)
Comment (code-elinka) [reply pkg/backends/backends_test.go]: changed, thank you

File: pkg/backends/backends_test.go [hunk 7]
```diff
@@ -48,6 +48,17 @@ func TestEnsureL4BackendService(t *testing.T) {
 			idleTimeout:                 defaultIdleTimeout,
 			trackingMode:                defaultTrackingMode,
 		},
+		{
+			desc:                        "Test case with disabled session affinity but prepared params",
```
Comment (mmamczur) [inline pkg/backends/backends_test.go:52]: I think "prepared params" is a but difficult to understand

what does it mean? that connection tracking policy is provided but the flag is false?

also, maybe we should have connectionTracking as item in the cases struct above and provide it filled when needed. We should test for nil case since it's the default and also for when it's set (for whatever reason) but the flag is off

**Comment by mmamczur** [comment]
looks ok, let's have @aojea  also look at it

File: pkg/backends/backends_test.go (hunk 7 above)
Comment (code-elinka) [reply pkg/backends/backends_test.go]: Can you check the test now, please?

File: pkg/loadbalancers/l4netlb.go [hunk 8]
```diff
@@ -308,8 +308,9 @@ func (l4netlb *L4NetLB) provideBackendService(syncResult *L4NetLBSyncResult, hcL
 		syncResult.GCEResourceInError = annotations.BackendServiceResource
 		syncResult.Error = err
 	}
+	needsConnectionTracking := connectionTrackingPolicy != nil
```
Comment (aojea) [inline pkg/loadbalancers/l4netlb.go:319]: ok, this is where we check if the flag is enabled`connectionTrackingPolicy, err := l4netlb.connectionTrackingPolicy()`

Since we return nil for `connectionTrackingPolicy` we can avoid to add an additional bool argument and use the nil value to know if the feature is enabled or disabled

File: pkg/backends/backends_test.go [hunk 9]
```diff
@@ -36,45 +36,74 @@ func TestEnsureL4BackendService(t *testing.T) {
...
+			schemeType:       string(cloud.SchemeInternal),
 		},
 		{
-			desc:                        "Test Backend Service with Strong Session Affinity configuration",
-			serviceName:                 "test-service",
-			serviceNamespace:            "test-ns",
-			protocol:                    "TCP",
-			affinityType:                string(v1.ServiceAffinityClientIP),
-			schemeType:                  string(cloud.SchemeExternal),
-			enableStrongSessionAffinity: true,
```
Comment (aojea) [inline pkg/backends/backends_test.go:58]: I don't see any test now with `enableStrongSessionAffinity` to true

File: pkg/backends/backends_test.go [hunk 10]
```diff
@@ -155,7 +184,7 @@ func TestEnsureL4BackendServiceDoesNotDetachBackends(t *testing.T) {
 		t.Fatalf("failed to create the existing backend service: %v", err)
 	}
 
-	bs, err := backendPool.EnsureL4BackendService(bsName, hcLink, "TCP", string(v1.ServiceAffinityNone), string(cloud.SchemeInternal), namespacedName, network, nil)
+	bs, err := backendPool.EnsureL4BackendService(bsName, hcLink, "TCP", string(v1.ServiceAffinityNone), string(cloud.SchemeInternal), namespacedName, network, false, nil)
```
Comment (aojea) [inline pkg/backends/backends_test.go:187]: you need to modify this test to run with this value to true and this valeue to false

```
for _, affinity := range []bool{true,false} {
   t.Run(  

}
```

File: pkg/backends/backends_test.go [hunk 11]
```diff
@@ -236,6 +265,32 @@ func TestBackendSvcEqual(t *testing.T) {
...
+			newBackendService: &composite.BackendService{
+				ConnectionTrackingPolicy: nil,
+			},
+			wantEqual: false,
+		},
+		{
+			desc:                      "Test to check if function ignores nil connection tracking policy",
+			compareConnectionTracking: false,
+			oldBackendService: &composite.BackendService{
+				ConnectionTrackingPolicy: nil,
```
Comment (aojea) [inline pkg/backends/backends_test.go:285]: the problematic is the empty value, add another test case with the empty value of `&composite.BackendServiceConnectionTrackingPolicy{}`

File: pkg/loadbalancers/l4netlb.go (hunk 8 above)
Comment (aojea) [reply pkg/loadbalancers/l4netlb.go:319]: but that is not needed ,just a comment, the bool sounds ok too

File: pkg/backends/backends_test.go (hunk 10 above)
Comment (code-elinka) [reply pkg/backends/backends_test.go]: Done

File: pkg/backends/backends_test.go (hunk 9 above)
Comment (code-elinka) [reply pkg/backends/backends_test.go:58]: oops, nice catch, thanks

File: pkg/loadbalancers/l4netlb.go (hunk 8 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go:317]: I removed double check for SSA flag (there was already a check in `l4netlb.connectionTrackingPolicy()`). PTAL. But I would not remove `needsConnectionTracking` because what if we had SSA enabled and the user would want to set `ConnectionTrackingPolicy` to `nil`.

File: pkg/loadbalancers/l4.go [hunk 12]
```diff
@@ -445,7 +448,7 @@ func (l4 *L4) EnsureInternalLoadBalancer(nodeNames []string, svc *corev1.Service
 	}
 
 	// ensure backend service
-	bs, err := l4.backendPool.EnsureL4BackendService(bsName, hcLink, string(protocol), string(l4.Service.Spec.SessionAffinity), string(cloud.SchemeInternal), l4.NamespacedName, l4.network, &composite.BackendServiceConnectionTrackingPolicy{})
+	bs, err := l4.backendPool.EnsureL4BackendService(bsName, hcLink, string(protocol), string(l4.Service.Spec.SessionAffinity), string(cloud.SchemeInternal), l4.NamespacedName, l4.network, false, nil)
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4.go:451]: regarding last 2 args:

- false will not be needed any more
- 'nil' should be repalced with named variable, defined in l4.go file
var  *composite.BackendServiceConnectionTrackingPolicy NoConnectionTrackingPolicy := nil

File: pkg/backends/backends.go [hunk 13]
```diff
@@ -288,7 +288,7 @@ func (b *Backends) DeleteSignedUrlKey(be *composite.BackendService, keyName stri
 }
 
 // EnsureL4BackendService creates or updates the backend service with the given name.
-func (b *Backends) EnsureL4BackendService(name, hcLink, protocol, sessionAffinity, scheme string, nm types.NamespacedName, network network.NetworkInfo, connectionTrackingPolicy *composite.BackendServiceConnectionTrackingPolicy) (*composite.BackendService, error) {
+func (b *Backends) EnsureL4BackendService(name, hcLink, protocol, sessionAffinity, scheme string, nm types.NamespacedName, network network.NetworkInfo, useConnectionTrackingPolicy bool, connectionTrackingPolicy *composite.BackendServiceConnectionTrackingPolicy) (*composite.BackendService, error) {
```
[the remark this answers is not in the document]
Comment (code-elinka) [reply pkg/backends/backends.go]: done

[the remark this answers is not in the document]
Comment (code-elinka) [reply pkg/backends/backends.go]: Probably done

File: pkg/loadbalancers/l4.go (hunk 12 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4.go]: I replaced it in internal controller but not in tests. Tell me if you need tests too.

Comment (cezarygerard) [reply pkg/loadbalancers/l4.go]: yes, you should use the same variable

File: pkg/loadbalancers/l4.go [hunk 14]
```diff
@@ -96,27 +97,29 @@ func NewL4ILBSyncResult(syncType string, startTime time.Time, svc *corev1.Servic
...
-	Namer            namer.L4ResourcesNamer
-	Recorder         record.EventRecorder
-	DualStackEnabled bool
-	NetworkResolver  network.Resolver
+	Service                      *corev1.Service
+	Cloud                        *gce.Cloud
+	Namer                        namer.L4ResourcesNamer
+	Recorder                     record.EventRecorder
+	DualStackEnabled             bool
+	StrongSessionAffinityEnabled bool // Flag to use SSA feature code, false if SSA code should be ignored
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4.go:105]: you don't need this for ILB

File: pkg/loadbalancers/l4.go [hunk 15]
```diff
@@ -96,27 +97,29 @@ func NewL4ILBSyncResult(syncType string, startTime time.Time, svc *corev1.Servic
...
-		networkResolver: params.NetworkResolver,
+		cloud:                              params.Cloud,
+		scope:                              scope,
+		namer:                              params.Namer,
+		recorder:                           params.Recorder,
+		Service:                            params.Service,
+		healthChecks:                       healthchecksl4.NewL4HealthChecks(params.Cloud, params.Recorder),
+		forwardingRules:                    forwardingrules.New(params.Cloud, meta.VersionGA, scope),
+		enableDualStack:                    params.DualStackEnabled,
+		enableStrongSessionAffinityFeature: params.StrongSessionAffinityEnabled,
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4.go:121]: you don't need this for ILB

File: pkg/loadbalancers/l4.go [hunk 16]
```diff
@@ -53,15 +53,16 @@ type L4 struct {
...
-	network         network.NetworkInfo
-	networkResolver network.Resolver
+	recorder                           record.EventRecorder
+	Service                            *corev1.Service
+	ServicePort                        utils.ServicePort
+	NamespacedName                     types.NamespacedName
+	forwardingRules                    ForwardingRulesProvider
+	healthChecks                       healthchecksl4.L4HealthChecks
+	enableDualStack                    bool
+	enableStrongSessionAffinityFeature bool // a flag for SSA feature to use SSA code
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4.go:63]: you don't need this for ILB

File: pkg/l4lb/l4controller.go [hunk 17]
```diff
@@ -246,12 +246,13 @@ func (l4c *L4Controller) processServiceCreateOrUpdate(service *v1.Service) *load
...
-		Namer:            l4c.namer,
-		Recorder:         l4c.ctx.Recorder(service.Namespace),
-		DualStackEnabled: l4c.enableDualStack,
-		NetworkResolver:  l4c.networkResolver,
+		Service:                      service,
+		Cloud:                        l4c.ctx.Cloud,
+		Namer:                        l4c.namer,
+		Recorder:                     l4c.ctx.Recorder(service.Namespace),
+		DualStackEnabled:             l4c.enableDualStack,
+		StrongSessionAffinityEnabled: l4c.ctx.EnableL4StrongSessionAffinity,
```
Comment (cezarygerard) [inline pkg/l4lb/l4controller.go:254]: you don't need this for ILB

File: pkg/l4lb/l4controller.go [hunk 18]
```diff
@@ -325,12 +326,13 @@ func (l4c *L4Controller) processServiceDeletion(key string, svc *v1.Service) *lo
...
-		Namer:            l4c.namer,
-		Recorder:         l4c.ctx.Recorder(svc.Namespace),
-		DualStackEnabled: l4c.enableDualStack,
-		NetworkResolver:  l4c.networkResolver,
+		Service:                      svc,
+		Cloud:                        l4c.ctx.Cloud,
+		Namer:                        l4c.namer,
+		Recorder:                     l4c.ctx.Recorder(svc.Namespace),
+		DualStackEnabled:             l4c.enableDualStack,
+		StrongSessionAffinityEnabled: l4c.ctx.EnableL4StrongSessionAffinity,
```
Comment (cezarygerard) [inline pkg/l4lb/l4controller.go:334]: you don't need this for ILB

File: pkg/loadbalancers/l4_test.go [hunk 19]
```diff
@@ -2161,12 +2161,13 @@ func mustSetupILBTestHandler(t *testing.T, svc *v1.Service, nodeNames []string)
...
-		Namer:            namer,
-		Recorder:         record.NewFakeRecorder(100),
-		DualStackEnabled: true,
-		NetworkResolver:  network.NewFakeResolver(network.DefaultNetwork(fakeGCE)),
+		Service:                      svc,
+		Cloud:                        fakeGCE,
+		Namer:                        namer,
+		Recorder:                     record.NewFakeRecorder(100),
+		DualStackEnabled:             true,
+		StrongSessionAffinityEnabled: false,
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4_test.go:2169]: you don't need this for ILB

**Comment by cezarygerard** [comment]
@Elina,

you need to update pkg/loadbalancers/l4netlb.go
in NewL4NetLB, line ~127 needs to use the new NewPoolWithConnectionTrackingPolicy

https://github.com/kubernetes/ingress-gce/blob/21745483f0365b0c4d717b2d2dc53bff517aaba4/pkg/loadbalancers/l4netlb.go#L127

File: pkg/loadbalancers/l4netlb.go [hunk 20]
```diff
@@ -207,15 +207,17 @@ func (l4netlb *L4NetLB) EnsureFrontend(nodeNames []string, svc *corev1.Service)
...
 	}
 	l4netlb.networkInfo = *networkInfo
-	// if service requires strong session affinity, check requirements
-	if err := l4netlb.checkStrongSessionAffinityRequirements(); err != nil {
-		result.Error = err
-		result.MetricsState.Status = metrics.StatusError
-		if utils.IsUserError(err) {
-			result.MetricsLegacyState.IsUserError = true
-			result.MetricsState.Status = metrics.StatusUserError
+	if l4netlb.enableStrongSessionAffinity {
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4netlb.go:210]: this is redundant

File: pkg/loadbalancers/l4netlb.go [hunk 21]
```diff
@@ -303,25 +305,29 @@ func (l4netlb *L4NetLB) provideBackendService(syncResult *L4NetLBSyncResult, hcL
 	servicePorts := l4netlb.Service.Spec.Ports
 	protocol := utils.GetProtocol(servicePorts)
 
-	connectionTrackingPolicy, err := l4netlb.connectionTrackingPolicy()
+	var bs *composite.BackendService
+	var err error
+	var connectionTrackingPolicy *composite.BackendServiceConnectionTrackingPolicy
+
+	connectionTrackingPolicy, err = l4netlb.connectionTrackingPolicy()
```
Comment (cezarygerard) [inline pkg/loadbalancers/l4netlb.go:312]: l4netlb.connectionTrackingPolicy() never returns error, skip the second return velue

File: pkg/loadbalancers/l4.go (hunk 14 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4.go]: done

File: pkg/loadbalancers/l4.go (hunk 16 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4.go]: done

File: pkg/l4lb/l4controller.go (hunk 17 above)
Comment (code-elinka) [reply pkg/l4lb/l4controller.go]: done

File: pkg/loadbalancers/l4.go (hunk 15 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4.go]: done

File: pkg/l4lb/l4controller.go (hunk 18 above)
Comment (code-elinka) [reply pkg/l4lb/l4controller.go]: done

File: pkg/loadbalancers/l4_test.go (hunk 19 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4_test.go]: done

File: pkg/loadbalancers/l4netlb.go [hunk 22]
```diff
@@ -185,7 +185,7 @@ func (l4netlb *L4NetLB) checkStrongSessionAffinityRequirements() *utils.UserErro
 // This function does not link instances to Backend Service.
 func (l4netlb *L4NetLB) EnsureFrontend(nodeNames []string, svc *corev1.Service) *L4NetLBSyncResult {
 	isMultinetService := l4netlb.networkResolver.IsMultinetService(svc)
-	enabledStrongSessionAffinity := annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
+	enabledStrongSessionAffinity := l4netlb.enableStrongSessionAffinity && annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
```
[the remark this answers is not in the document]
Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go]: done

File: pkg/loadbalancers/l4netlb.go [hunk 23]
```diff
@@ -411,8 +417,10 @@ func (l4netlb *L4NetLB) ensureIPv4NodesFirewall(nodeNames []string, ipAddress st
 // It is health check, firewall rules and backend service
 func (l4netlb *L4NetLB) EnsureLoadBalancerDeleted(svc *corev1.Service) *L4NetLBSyncResult {
 	isMultinetService := l4netlb.networkResolver.IsMultinetService(svc)
-	enabledStrongSessionAffinity := annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
-	result := NewL4SyncResult(SyncTypeDelete, svc, isMultinetService, enabledStrongSessionAffinity)
+	var result *L4NetLBSyncResult
```
[the remark this answers is not in the document]
Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go]: is it better now?

File: pkg/loadbalancers/l4netlb.go [hunk 24]
```diff
@@ -411,8 +417,10 @@ func (l4netlb *L4NetLB) ensureIPv4NodesFirewall(nodeNames []string, ipAddress st
...
+	enabledStrongSessionAffinity := l4netlb.enableStrongSessionAffinity && annotations.HasStrongSessionAffinityAnnotation(l4netlb.Service)
```
[the remark this answers is not in the document]
Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go]: done

File: pkg/loadbalancers/l4netlb.go (hunk 21 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go]: removed err from return values

File: pkg/loadbalancers/l4.go (hunk 12 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4.go]: done, PTAL

File: pkg/loadbalancers/l4netlb.go (hunk 20 above)
Comment (code-elinka) [reply pkg/loadbalancers/l4netlb.go]: done

**Comment by mmamczur** [comment]
looks ok after changes
